

# Modeling evolution: idealized populations


### Interactive Notebook
...


## Learning objectives
By the end of this lesson you should:

- Know which assumptions underlie the definition of an 'idealized population'
- Understand why allele frequencies do not change in the absence of evolutionary processes.
- Be able to calculate genotype frequencies under Hardy-Weinberg equilibrium given allele frequencies at a locus.
- Understand why violations of the Hardy-Weinberg assumptions cause allele frequencies to change over generations.


## Genetic variation
Consider a population that has variation at a single gene (A) in the form
of two alleles (A1 and A2) which differ by a single mutation. We may be 
interested in the frequency of the two alleles in this population; how 
much more frequent does one occur relative to the other? If we assume our
diploid population is of size N, then there are 2N gene copies, every one 
of which exists as either an A1 or A2 *allele*. How can we best measure 
the genetic diversity of this population? How many samples do we need? And 
what can we learn about a population by measuring its genetic diversity at 
a single gene? These types of questions are typically considered to be 
within the discipline of *population genetics*, a field that has traditionally
been treated as distinct from *phylogenetics*. However, as we will see, 
developments in recent decades have significantly bridged a gap between
these fields. Major questions in both disciplines, such as predicting
changes in allele frequencies, or estimating divergence times between 
species, can both be modeled as an outcome of genealogical (tree-based)
processes.

The term *allele* above refers to the state of information in a gene 
copy, i.e., its *haplotype*, which arose as a consequence of mutations 
throughout its genealogical history. So far, we are considering only
two alleles in this population. If a new mutation arose at a single 
gene copy of this gene, then we could also consider a third allele, 
such as A3; but for simplicity, we will assume for now that only 
two alleles exist in the population. *In fact, these types of 
simplifying assumptions are the topic of this lesson.*

## Genetic variation and genealogies
Throughout this chapter we will reinforce a clear distinction between 
relatedness of gene copies, which is a reflection of their 
*usually unobservable* genealogical relationships, versus the 
identity of gene copies, such as their allelic state, which is an 
*observable/measurable* character that arose as a consequence of 
mutations occurring throughout their genealogical history.
<!-- and which can be used to infer their genealogy.  -->

Our example above is a clear case of this distinction. All 2N gene copies
share a genealogy in which some are more closely related to others, however,
because there is only a single mutation in the population, there is at most 
enough information to group these gene copies into two groups, and certainly
not enough information to infer their full genealogy. <!-- This topic will 
be reinforced in the coming chapters through the use of coalescent 
simulation software. -->

## Population genetics
Let's return to our example of genetic variation in a population. A primary
goal of population genetics is to develop models of evolution within
populations that can be used to predict changes in allele frequencies 
over time. Such models can be discrete or continuous in time, and can 
include parameters to describe how individuals live, die, reproduce, and 
interact with each other and their environment. However, as we will see,
there is a benefit to constructing simple mathematical models to describe
evolution in populations, models that include as few of processes and 
parameters as possible, while still being informative. 

In many cases, these simple models can be incredibly powerful for 
describing evolution. Here we will focus on two genealogical models
of evolution within populations, the **Wright-Fisher (WF) process**, and 
the **Coalescent**. The latter is a mathematical approximation of the
former, and so we will first focus on the WF process. 

<!-- The connection 
between these simple "bean bag" models of population genetics, and 
-->




<!--  
### Model assumptions
Before proceeding to the topic of coalescent models and simulation 
techniques, however, it is important to understand fundamental assumptions
underlying these models. For example, 
 -->

### Idealized populations
Population genetics is the study of allele frequencies in populations. 
Amazingly, much of the foundational work in this field was developed 
*even before we knew about DNA*, and involved only combining probability
and statistics with Mendel's theory of [particulate (non-blending) 
inheritance](https://en.wikipedia.org/wiki/Mendelian_inheritance). 

Mendel showed that genes are discrete units, diploid organisms contain 
two copies at every gene, and that the two copies have equal probability 
of being passed on to offspring through the formation of haploid gametes. 
This simple law of *segregation and independent assortment* can be 
formalized as a probability statement, and used to develop probabilistic 
models.

For this, **we need to make some assumptions about populations.** 
An *idealized population* refers to a theoretical population that meets 
a number of unrealistic but useful assumptions. Most notably, that it is 
of *infinite size* and that the individuals within it are *randomly mating*.
As we will see, these and other assumptions can be relaxed to allow for 
further insights into their effect on the model. Overall, the purpose of
these population genetic models is to understand how and why allele 
frequencies (i.e., the relative abundance of A1 versus A1) change 
over time.


### Hardy-Weinberg Expectation
The [*Hardy-Weinberg* equilibrium](https://en.wikipedia.org/wiki/Hardy%E2%80%93Weinberg_principle) is a statement that **the frequencies of 
alleles (variants at a gene) and genotypes (combinations of alleles at a 
gene) will remain constant** through time in an *idealized population* 
in the absence of <span style="color:red">selection, mutation, migration, 
and genetic drift.</span> 

This is not actually a super surprising statement in terms of statistics, 
but more so in terms of evolution. It is identifying that 
<span style="color:red">these processes are fundamental to explaining why 
allele frequencies change over time</span>, and also, that the process of 
*segregation of alleles* during Meiosis (i.e., Mendelian Inheritance) is 
not expected to change allele frequencies on its own. 
The concept of Hardy Weinberg equilibrium is demonstrated in the figure below.

==insert figure here==

<!-- <img src=https://eaton-lab.org/slides/fundamentals2019/session-9-popgen/data/hardy-weinberg.png style="width:60%"> -->

In the first cell of the figure above the population is initially not in 
Hardy-Weinberg equilibrium. We can tell this by looking at the genotype 
frequencies. There appears to be a deficit of heterozygotes. But what is the 
null expectation for the number of heterozygotes that should exist? Well, this
is something that Hardy-Weinberg can tell us. To answer that question we need
to calculate the genotype frequencies after just one generation of random 
mating in an infinite-sized population starting with the allele frequencies
that exist in the population currently.

After a single generation genotype frequencies (e.g., A1A1, A1A2, and A2A2) 
will reach HW equilibrium, despite the fact that allele frequencies (e.g., 
A1 and A2) will not change. This expectation can be computed exactly using 
probability (here I will use the same genotype frequencies as in the figure 
above, but change the allele names to A and B, rather than A1 and A2 to make 
it easier to type).

```python
import numpy as np
```

```python title="initial diploid genotype frequencies"
AA = 0.3
AB = 0.0
BB = 0.7
```

Each AA parent produces two A alleles, each AB parent produces one A and 
one B allele, and each BB parent produces two B alleles. Based on this 
simple fact we can calculate the relative frequency of A and B alleles, 
which we label p and q. For example, p is equal to the frequency of AA 
homozygotes plus 1/2 the frequency of heterozygotes; q is equal to the 
frequency of BB homozygotes plus 1/2 the frequency of heterozygoes.

You might ask, but what if some of the diploids produced more alleles than
others? Well, for that to happen would be a manifestation of either drift 
or selection, both of which we are assuming to be absent from this model. 
This is key to remember.


```python title="haploid gametes generated by diploids (given model assumptions)"
p = AA + (AB / 2)
q = BB + (AB / 2)
print(p, q)  # 0.3 0.7
```

Now that we know the frequency of alleles in the gametes we can calculate 
the frequency of genotypes formed in the next generation by assuming that 
populations are randomly mating. AA genotypes will occur with probability 
of sampling p twice (p * p), heterozygotes are the probability of sampling 
p and q (or q and p), and BB homozygotes are the probability of sampling q 
twice (q * q). Once again, note the importance of the assumptions underlying
our idealized population for this model prediction.
<!-- HW expectation: `p**2 + 2pq + q**2 = 1` -->

```python title="Genotype frequencies of *next* generation of diploids."
newAA = p * p
newAB = 2 * p * q
newBB = q * q
print(newAA, newAB, newBB)  # 0.09 0.42 0.48999999999999994
```


```python title="Generate bar plots."
import toyplot

canvas = toyplot.Canvas(width=400, height=400)
ax0 = canvas.cartesian(grid=(0, 2, 2))
ax0.bars()
# see notebook in progress. tuba:~/Documents/2020-fundamentals/notebooks
```









### Why HW is relevant to phylogenetics
HW is a model for making predictions about allele frequencies in an
*idealized population*. 
