---
---

# Wright-Fisher

## Learning objectives
By the end of this lesson you will be familiar with:

1. The Wright-Fisher (WF) process model.
2. Assumptions underlying this model.
3. calculating the probability two gene copies share a parent.
4. extensions of this probabilistic thinking to studying genealogies.


## Wright-Fisher Process
The Wright-Fisher process is considered one of the simplest models in 
population genetics. Compared to an idealized population, in which 
allele frequencies are always in HW equilibrium, and do not change, 
the WF process by default relaxes at least one assumption: population 
sizes are not infinite. 

The introduction of finite population sizes allows for allele frequencies
to change over time, as genetic drift will cause an allele to become either 
entirely fixed, or lost (i.e., reaches frequency 1.0 or 0.0). Consequently,
genetic variation will be lost over time, with the rate of loss being 
a function of the population size. 


## Effective population size
One of the most challenging concepts in population genetics is that of 
*effective population size* ($N_e$). This is not to be confused with
the population census size, which is the number of individuals in a 
population. By contrast, $N_e$ is an emergent property of a population 
that is a consequence of many possible ways in which it does not fit the 
assumptions of an idealized population. Examples include finite population 
census sizes, non-random mating, mutation, selection, and population 
structure and migration.

Because in the standard WF process the only way in which it differs from
an idealized population is by having a finite population size, the WF process
provides a convenient way to study the effects of $N_e$, since in this case, 
and this case alone, $N_e$ will be equal to the population census size. 
Thus, we can examine how raising or lowering $N_e$ affects the rate of 
fixation of alleles. 


## Visualizing the WF process
To help understand WF models and their connection to genealogies it is
helpful to visualize stochastic outcomes of a WF process. The following
visualizations are created using `toyplot`, and you can learn more about
the visualization code in the next [code lesson]. Here we will simply use
the `WrightFisherPlot` class from the dropdown below to generate and draw
WF process outcomes. This takes a `seed` parameter for a random number
generator to produce different outcomes. 

??? example "WF process visualization code using toyplot"
	
	```py
	class WrightFisherPlot:
	    def __init__(self, time=20, popsize=20, width: int=500, height: int=500, seed: Optional[int]=None):
	        self.rng: np.random.Generator = np.random.default_rng(seed)
	        self.grid: np.ndarray = np.arange(popsize * 2 * time).reshape((time, popsize * 2))
	        self.edges: np.ndarray = None
	        self.canvas: toyplot.Canvas = toyplot.Canvas(width=width, height=height)
	        self.axes: toyplot.coordinates.Cartesian = None
	        self._setup_axes()
	        self.marks: Dict[str, 'toyplot.Mark'] = {}

	        # get node coordinates
	        ys, xs = np.where(self.grid > -1)
	        self.coords = np.column_stack([xs, ys])

	    def _setup_axes(self):
	        # setup the canvas and axes
	        self.axes = self.canvas.cartesian(xshow=False, ylabel="Time (generations)", margin=70, padding=25)
	        self.axes.y.ticks.labels.style["font-size"] = 14
	        self.axes.y.label.style["font-size"] = 16
	        self.axes.y.label.offset = 30
	        self.axes.y.ticks.locator = toyplot.locator.Extended()
	        self.axes.y.ticks.show = True
	        
	    def add_haploids(self, **kwargs):
	        """Add circle marks for gene copies."""
	        style = {
	            "marker": "o", 
	            "color": "white", #toyplot.color.Palette()[1]
	            "size": self.canvas.width / self.grid.shape[1] / 7.5,
	            "mstyle": {"stroke": "black", "stroke-opacity": 0.75, "stroke-width": 1.5},
	        }
	        style.update(kwargs)
	        self.marks['haploids'] = self.axes.scatterplot(
	            self.coords[:, 0], self.coords[:, 1], 
	            **style,
	        );

	    def add_diploids(self):
	        """Adds a rectangle around pairs of gene copies to represent a diploid individual."""
	        self.marks['diploids'] = self.axes.rectangle(
	            self.coords[:, 0][::2] - 0.25,
	            self.coords[:, 0][::2] + 1.25, 
	            self.coords[:, 1][::2] - 0.25,
	            self.coords[:, 1][::2] + 0.25,
	            style={"fill": "lightgrey", "stroke": "grey", "stroke-opacity": 0.75, "stroke-width": 1.5},
	        );
	        
	    def add_lines(self, sort: bool=True, **kwargs):
	        """Adds lines from gene copies to randomly sampled parent gene copies each generation.
	        """
	        # iterate over each generation adding pairs of node indices
	        for gen in range(0, self.grid.shape[0] - 1):
	            # children idxs span from left to right
	            lower_idxs = self.grid[gen]

	            # randomly sample parent idxs (some have many children, some none)
	            upper_idxs = self.rng.choice(self.grid[gen + 1], size=self.grid[gen + 1].size, replace=True)
	            
	            # get sorting index for the upper idxs
	            if sort:
	                order = np.argsort(upper_idxs)
	            else:
	                order = np.arange(upper_idxs.size)
	                
	            # update array of edges
	            iedges = np.column_stack([lower_idxs, upper_idxs[order]])
	            if self.edges is not None:
	                self.edges = np.concatenate([self.edges, iedges])
	            else:
	                self.edges = iedges

	        # style the graph
	        style = {"vlshow": False, "ecolor": "black", "ewidth": 1.5, 'vsize': 0, 'estyle': {}}
	        style.update(kwargs)

	        # add graph lines from lower_idxs to upper_idxs, using coordinates for all
	        self.marks['genealogy'] = self.axes.graph(
	            self.edges,
	            vcoordinates=self.coords[sorted(np.unique(self.edges))],
	            **style,
	        );
	        
	    def add_sampled_lines(self, nsample: Optional[int]=None, **kwargs):
	        """Add sampled genealogy lines."""        
	        if not nsample:
	            return None
	        
	        # style the graph
	        style = {"vlshow": False, "ecolor": toyplot.color.Palette()[0], "ewidth": 2, 'vsize': 0, 'estyle': {}}
	        style.update(kwargs)
	            
	        # color selected edges
	        nsample = min(nsample, self.grid.shape[0])
	        tracked = list(self.rng.choice(self.grid[0], size=nsample, replace=False))
	        for e in self.edges:
	            if e[0] in tracked:
	                tracked.append(e[1])

	        mask = np.isin(self.edges[:, 0], tracked)
	        edges = self.edges[mask]

	        # add graph lines from lower_idxs to upper_idxs, using coordinates for all
	        self.marks['genealogy'] = self.axes.graph(
	            edges,
	            vcoordinates=self.coords[sorted(np.unique(edges))],
	            **style,
	        );
	```


### Simulating WF process

A WF process can be simulated quite easily. It is a discrete-time process
where each generation 2N haploid gene copies exist. Because the population
is assumed to be randomly mating, we can randomly assign haploid gene copies
to be grouped into diploid individuals. This grouping has no effect on how
the model operates (unless some other model assumptions were to be relaxed) 
and is only for visualization purposes. To simulate one generation of the 
WF process one needs only to create a new generation of 2N haploid copies
and to randomly 2N copies, with replacement, from the previous generation
to serve as their parents. 

The result of this random mating process, played out over multiple 
generations, looks like the plot below. The process gives rise to a 
genealogy of gene copies. Lines are drawn to connect gene copies in each
generation from parent to offspring. Grey rectangles represent diploid 
individuals, which contain two gene copies, represented by red circles.

```py
# example WF plot
wf = WrightFisherPlot(seed=123, time=11, popsize=10, width=700, height=500)
wf.add_diploids()
wf.add_lines(sort=False)
wf.add_haploids(color="red", size=8)
```

<figure markdown>
  ![WF-plot-2.2.0](../chapter-1/notebooks/fig.2.2.0.svg){width="700", loading=lazy}
  <figcaption>Fig. 2.2.0. Wright-Fisher process.</figcaption>
</figure>

### Interpreting WF genealogies

This process is much easier to comprehend when the edges are sorted so that
they are easier to read. Remember, however, that regardless of this sorting
for visualization purposes, that the parents of gene copies are randomly 
sampled each generation. Time in this model flows from top to bottom, such
that the gene copies at time=0 represent the present, and the ancestors
of these gene copies, as well as other gene copies that did not leave 
descendants, exist in past generations (gens 1 - 11 in this visualization).


```py
# example WF plot with sorted edges for visualization
wf = WrightFisherPlot(seed=123, time=11, popsize=10, width=700, height=500)
wf.add_diploids()
wf.add_lines(sort=True)
wf.add_haploids(color="red", size=8)
```

<figure markdown>
  ![WF-plot-2.2.1](../chapter-1/notebooks/fig.2.2.1.svg){width="700", loading=lazy}
  <figcaption>Fig. 2.2.1. WF process with edges sorted for visualization.</figcaption>
</figure>

Understanding the flow of time in these diagrams is very important, and 
can sometimes be challenging, since different authors may present the same
type of figure flowing top to bottom, or bottom to top. Fortunately, there
is a simple trick to easily identify the flow of time, which is to identify
the *direction in which every node has an edge flowing*. That direction will
be backwards in time. This is because **every gene copy has an ancestor, 
but not every gene copy will leave descendants**. The latter outcome is 
a consequence of genetic drift, where a random sampling process will cause
some gene copies to leave multiple descendants, while others will leave none.

The plot below zooms in on just the final three generations of a WF process
simulation. You can see that every gene copy has an edge pointing up
(backwards in time), but many gene copies do not have an edge pointing down.

As we said earlier, mating occurs randomly in this model, without regard to
the way in which gene copies are grouped inside of diploid individuals. 
Consequently, it is possible for odd mating strategies to occur, such as 
selfing, which is a case where a diploid individual produces an offspring
genetically identical to itself. In the figure below, generated with the
random seed 123, look at the first two individuals from the left in generation
0. The first inherits the same two gene copies that present in an individual
in the previous generation. The second inherits two copies of the same gene
copy that was present in an individual in the previous generation. The first
is a representation of diploid selfing, and the second a form of haploid 
selfing. 

```py
# example WF plot with sorted edges for visualization
wf = WrightFisherPlot(seed=123, time=3, popsize=10, width=700, height=300)
wf.add_diploids()
wf.add_lines(sort=True)
wf.add_haploids(color="red", size=8)
```

<figure markdown>
  ![WF-plot-2.2.2](../chapter-1/notebooks/fig.2.2.2.svg){width="700", loading=lazy}
  <!-- <figcaption>An example phylogenetic tree drawing.</figcaption> -->
</figure>


## Genealogy of samples
A consequence of this simple discrete random sampling process, repeated 
over multiple generations, is that any gene copy that exists at the 
present can trace back a tree-like genealogical history with any other 
gene copy at the present. As an example, we can randomly sample 5 gene 
copies at the present and focus on their individual ancestries. As we 
can see, they each trace back some number of generations before they 
eventually share a common ancestor; an ancestral gene copy from which 
they are all descended.

```py
# example WF plot with sampled tips genealogy
wf = WrightFisherPlot(seed=123, time=21, popsize=10, width=700, height=600)
wf.add_lines(sort=True, eopacity=0.25)
wf.add_sampled_lines(nsample=5, ecolor="black", ewidth=2.5, vsize=10, vcolor='black')
wf.add_haploids(color="black", size=5, opacity=0.25)
```

<figure markdown>
  ![WF-plot-2.2.3](../chapter-1/notebooks/fig.2.2.3.svg){width="700", loading=lazy}
  <!-- <figcaption>An example phylogenetic tree drawing.</figcaption> -->
</figure>




## Implementing Wright-Fisher

- analytical solutions for very simple models.
- simulations for more complex models
	- forward-in-time simulations with SLiM3. 
	- backwards-in-time (coalescent) simulations with msprime.

## Summary
