
# Introduction to Computational Phylogenetics


### Learning objectives
This chapter will introduce the topic of trees as data objects, and what this
means.


### Computational phylogenetics

<!-- Phylogenetics is what? And is defined as what? -->
The term *phylogenetics* refers to a broad array of scientific methods used
... and has evolved over time. It is sometimes viewed as separate from 
population genetics, which is concerned with changes over shorter evolutionary
time scales. Despite major differences in scope, and many common assumptions, 
however, this distinction is mostly artificial. All evolutionary processes
can be contextualized through the study of genealogical (tree-based)

<!-- Why is it so computational? -->
- statistical
- heuristic
- algorithmic, Bayesian, machine learning, linear algebra, calculus.

<!-- Why is it so computational? -->


### What is a phylogenetic tree?



### What does a tree represent?
A tree represents a hierarchical relationship. In the context of evolutionary
biology these relationships are typically interpreted as a pattern of ancestry
through time, i.e., ancestor-descendant relationships. 


#### Ancestor-descendant relationships
- genes and gene families
- individuals (genealogical)
- haplotypes (genealogical)
- population or species (model)
	- what is a species? A concept? Then how can we model the relationships
	among conceptual things that we can't define precisely?


However, this is not always the case, and thus to properly understand a 
tree it is important to understand what underlying data was used to infer
the tree, and what type of inference method was used.



#### Arbitrary representations of similarity
Trees are used widely throughout mathematics, statistics, biology, and many
other fields to represent any type of data that has a hierarchical structure, 
which includes pretty much any quantitative data for which pairwise distances
can be measured. 

Such approaches are employed commonly in the social sciences to group samples
by some quantitative measure of similarity. Examples include the study of 
human language families^1, of belief systems^2. Phylogenetic methods are
thus not solely applicable to inferring evolutionary patterns. 

You will often see results displaying a figure that looks something like 
the one below. Let's consider an example like that from ==[this study] by 
XXX et al. (xxxx)==, which used transcriptome sequencing to measure gene 
expression of different tissue types in a mouse and then performed 
UPGMA clustering to ...

Chapter ==XXX== covers distance-based phylogenetic inference methods
in much more detail.


??? Example "Example code to create hierarchical tree from random data points"
	```py
	import toytree
	import numpy as np
	from scipy import spatial, cluster

	# init a random number generator
	rng = np.random.default_rng(seed=123)

	# generate 100 random data points for 12 'samples'
	data = rng.random(size=(12, 100))

	# calculate a measure of distance between samples
	dists = toytree.distance.calculate_distances(data, method="pdist")

	# infer a tree by clustering samples by avg distance
	edges = toytree.distance.cluster_distances(dists, method="average")

	# translate edges into a tree
	tree = toytree.distance.get_tree_from_dists(edges)

	# draw the tree
	canvas, axes, mark = tree.draw()

	# draw a heatmap next to the tree
	mark2 = axes.table(data)
	```


<figure markdown>
  ![Heatmap image](../images/1.1.1-heatmap.png){ width="400", loading=lazy }
  <figcaption>A tree representing hierarchical relationships among 	
  	arbitrary data looks similar to an evolutionary tree, but has 
	a different meaning.
  </figcaption>
</figure>


### Evolutionary models in phylogenetics
How are evolutionary models distinct from other tree-based methods? 
Evolutionary models are probabilistic 




### References

- https://github.com/mam288/Bioinformatics-IV